src/CommonRuntime/Caching.fs

```diff
@@ -19,12 +19,16 @@ let createNonCachingCache() =
       member __.Set(_, _) = ()
       member __.TryRetrieve(_) = None }
 
+
 /// Creates a cache that uses in-memory collection
-let createInMemoryCache expiration = 
+let createInMemoryCache (expiration:TimeSpan) = 
   let dict = new ConcurrentDictionary<_, _>()
   { new ICache<_> with
       member __.Set(key, value) = 
         dict.[key] <- (value, DateTime.UtcNow)
+        async { do! Async.Sleep (expiration.Milliseconds) 
+                if dict <> null then 
+                    dict.TryRemove(key) |> ignore } |> Async.Start 
```

Comment:
This has a problem, if I add value A for key K, then after 5 seconds I set value B for key K, K will still be invalidated

Reply:
Isn’t that the desired behaviour thou. Invalidation should trump all shouldn’t it? Otherwise things can/will get out IFB sync leading to weird errors.

Reply:
You are absolutely correct. (I just copied the caching pattern that was used other places in this provider.) However, invalidating too early only leads to worst-case-scenario of doing a non-cached call twice. Instead of current huge memory leaks, which makes the whole XMLProvider unusable and Visual Studio busy.